SU2_DEF/src/drivers/CDiscAdjDeformationDriver.cpp

```diff
@@ -0,0 +1,945 @@
+          if (config->GetMarker_All_DV(iMarker) == YES) {
+            for (iVertex = 0; iVertex < geometry->nVertex[iMarker]; iVertex++) {
+
+              iPoint = geometry->vertex[iMarker][iVertex]->GetNode();
+              if ((iPoint < geometry->GetnPointDomain()) && UpdatePoint[iPoint]) {
+
+                Normal = geometry->vertex[iMarker][iVertex]->GetNormal();
+                VarCoord = geometry->vertex[iMarker][iVertex]->GetVarCoord();
+                Sensitivity = geometry->vertex[iMarker][iVertex]->GetAuxVar();
+
+                dS = 0.0;
+                for (iDim = 0; iDim < geometry->GetnDim(); iDim++) {
+                  dS += Normal[iDim]*Normal[iDim];
+                  deps[iDim] = VarCoord[iDim] / delta_eps;
+                }
+                dS = sqrt(dS);
+
+                dalpha_deps = 0.0;
+                for (iDim = 0; iDim < geometry->GetnDim(); iDim++) {
+                  dalpha[iDim] = Normal[iDim] / dS;
+                  dalpha_deps -= dalpha[iDim]*deps[iDim];
+                }
+
+                my_Gradient += Sensitivity*dalpha_deps;
+                UpdatePoint[iPoint] = false;
+              }
+            }
+          }
+        }
+      }
+
+      SU2_MPI::Allreduce(&my_Gradient, &localGradient, 1, MPI_DOUBLE, MPI_SUM, SU2_MPI::GetComm());
+      Gradient[iDV][0] += localGradient;
+    }
+  }
+
+  /*--- Delete memory for parameterization. ---*/
+
+  if (FFDBox != nullptr) {
+    for (iFFDBox = 0; iFFDBox < MAX_NUMBER_FFD; iFFDBox++) {
+      if (FFDBox[iFFDBox] != nullptr) {
+        delete FFDBox[iFFDBox];
+      }
+    }
+    delete [] FFDBox;
+  }
+
+  delete [] UpdatePoint;
+
+}
+
+
+void CDiscAdjDeformationDriver::SetDiscAdjDeformation_AD(CGeometry *geometry, CConfig *config, CSurfaceMovement *surface_movement, su2double** Gradient){
+
+  su2double DV_Value, *VarCoord, Sensitivity, my_Gradient, localGradient, *Normal, Area = 0.0;
+  unsigned short iDV_Value = 0, iMarker, nMarker, iDim, nDim, iDV, nDV, nDV_Value;
+  unsigned long iVertex, nVertex, iPoint;
+
+  int rank = SU2_MPI::GetRank();
+
+  nMarker = config->GetnMarker_All();
+  nDim    = geometry->GetnDim();
+  nDV     = config->GetnDV();
+
+  VarCoord = nullptr;
+
+  /*--- Discrete adjoint gradient computation ---*/
+
+  if (rank == MASTER_NODE)
+    cout  << endl << "Evaluate functional gradient using Algorithmic Differentiation (ZONE " << config->GetiZone() << ")." << endl;
+
+  /*--- Start recording of operations ---*/
+
+  AD::StartRecording();
+
+  /*--- Register design variables as input and set them to zero
+   * (since we want to have the derivative at alpha = 0, i.e. for the current design) ---*/
+
+
+
+  for (iDV = 0; iDV < nDV; iDV++){
+
+    nDV_Value =  config->GetnDV_Value(iDV);
+
+    for (iDV_Value = 0; iDV_Value < nDV_Value; iDV_Value++){
+
+      /*--- Initilization with su2double resets the index ---*/
+
+      DV_Value = 0.0;
+
+      AD::RegisterInput(DV_Value);
```

Comment:
You will need to apply some of the recent changes I made to SU2_DOT as this does not work anymore (it was a logic error for AD to do it like this).